### PR TITLE
Override browser key autorepeat with a custom class

### PR DESCRIPTION
**Pull request by frostburn**

### CODE REVIEW

File: ui/fancy/src/js/autorepeat.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+class KeyAutorepeater {
+    constructor(repeatDelay, repeatRate) {
+      this.repeatDelay = repeatDelay;
+      this.repeatRate = repeatRate;
+      this.listeners = new Map();
+      this.repeaters = new Set();
+      this.keyDownData = new Map();
+      window.addEventListener('keydown', (ev) => {
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: You really want to install these listeners in constructor? There should be separate method for it.

File: ui/fancy/src/js/autorepeat.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if (!this.listeners.has(ev.key)) {
+          return;
+        }
+        ev.preventDefault();
+        if (this.keyDownData.has(ev.key)) {
+          return;
+        }
+        let id = null;
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: This should be on first line of the function. And there is no need to set it `null`.

File: ui/fancy/src/js/autorepeat.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        if (this.repeaters.has(ev.key)) {
+          id = window.setTimeout(() => this.autorepeat(ev.key), this.repeatDelay);
+        }
+        this.keyDownData.set(ev.key, {
+          key: ev.key,
+          id: id,
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: `id` is enough.

File: ui/fancy/src/js/autorepeat.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        });
+        this.listeners.get(ev.key)();
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: This will bork if `this.listeners.get(ev.key)` returns undefined.

File: ui/fancy/src/js/autorepeat.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      });
+      window.addEventListener('keyup', (ev) => {
+        if (!this.keyDownData.has(ev.key)) {
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: Or you could just
```JavaScript
const data = this.keyDownData.get(ev.key);

if (data) {
   ...
}
```

File: ui/fancy/src/js/autorepeat.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        const data = this.keyDownData.get(ev.key);
+
+        window.clearTimeout(data.id);
+        this.keyDownData.delete(ev.key);
+      });
+    }
+
+    autorepeat(key) {
+      const id = window.setTimeout(() => this.autorepeat(key), this.repeatRate);
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: Missing blank line after local variable declaration.

File: ui/fancy/src/js/autorepeat.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      this.keyDownData.get(key).id = id;
+      this.listeners.get(key)();
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: Expects listeners to contain `key`.

File: ui/fancy/src/js/autorepeat.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
```
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: Too many blank lines.

File: ui/fancy/src/js/autorepeat.js (hunk 4 above)
Comment (frostburn) [inline ui/fancy/src/js/autorepeat.js]: The `.has` has been checked before this line.

File: ui/fancy/src/js/autorepeat.js (hunk 7 above)
Comment (frostburn) [inline ui/fancy/src/js/autorepeat.js]: Which is a valid expectation. (See `.has` comment before.)

File: ui/fancy/src/js/autorepeat.js (hunk 4 above)
Comment (RauliL) [inline ui/fancy/src/js/autorepeat.js]: You should use `.get` instead and check whether it returns `undefined`.

**Comment by frostburn** [comment]
Issues dealt with.

**Pull request merged**